#### 1. Test Skills Matcher API GET request using the default request JSON from API Explorer
#### 2. Submit custom survey answers

---

In [2]:
import pandas as pd
import requests 
from bs4 import BeautifulSoup
import yaml
import os
import json


/Users/alexsocarras/tech_impact/career-onestop/venv/lib/python3.9/site-packages/urllib3/__init__.py:34: NotOpenSSLWarning: urllib3 v2.0 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


#### 1. Test Skills Matcher API GET request using the default request JSON from API Explorer

The API explorer provides an example request to the skillsmatcher endpoint which we use now to test our first GET request through Python.
You can also submit requests to this endpoint via the API explorer -- the response you'll get there matches this one.

In [3]:
with open("api-key.yaml", "r") as file:
  api_key = yaml.full_load(file)

url = f"https://api.careeronestop.org/v1/skillsmatcher/{api_key['cs']['user-id']}"

with open(os.path.join("data","skills-submit-api", "example-request.json"), "r") as file:
  request_body = json.load(file)

headers = {
  "Authorization": f"Bearer {api_key['cs']['token-key']}"
}

In [6]:

response = requests.post(url, json=request_body, headers=headers)
response = response.json() # This matches the response we generated using the API Explorer

response

{'SKARankList': [{'OnetCode': '29-1229.04',
   'Score': 0.579203491091142,
   'Rank': 1,
   'Outlook': 'Below Average',
   'AnnualWages': 223410.0,
   'TypicalEducation': 'Doctoral or professional degree',
   'OccupationTitle': 'Physical Medicine and Rehabilitation Physicians',
   'EduCode': 2.0},
  {'OnetCode': '29-1141.02',
   'Score': 0.5451239172983321,
   'Rank': 2,
   'Outlook': 'Bright',
   'AnnualWages': 81220.0,
   'TypicalEducation': "Bachelor's degree",
   'OccupationTitle': 'Advanced Practice Psychiatric Nurses',
   'EduCode': 3.0},
  {'OnetCode': '29-1299.01',
   'Score': 0.5427063472592409,
   'Rank': 3,
   'Outlook': 'Below Average',
   'AnnualWages': 106230.0,
   'TypicalEducation': "Master's degree",
   'OccupationTitle': 'Naturopathic Physicians',
   'EduCode': 2.0},
  {'OnetCode': '29-1291.00',
   'Score': 0.5307120733164209,
   'Rank': 4,
   'Outlook': 'Average',
   'AnnualWages': 72220.0,
   'TypicalEducation': "Master's degree",
   'OccupationTitle': 'Acupuncturis

In [7]:

headers = {
  "Authorization": f"Bearer {api_key['token-key']}"
}

response = requests.post(url, json=example_request, headers=headers)
response = response.json() # This matches the response we generated using the API Explorer

response

{'SKARankList': [{'OnetCode': '29-1229.04',
   'Score': 0.579203491091142,
   'Rank': 1,
   'Outlook': 'Below Average',
   'AnnualWages': 223410.0,
   'TypicalEducation': 'Doctoral or professional degree',
   'OccupationTitle': 'Physical Medicine and Rehabilitation Physicians',
   'EduCode': 2.0},
  {'OnetCode': '29-1141.02',
   'Score': 0.5451239172983321,
   'Rank': 2,
   'Outlook': 'Bright',
   'AnnualWages': 81220.0,
   'TypicalEducation': "Bachelor's degree",
   'OccupationTitle': 'Advanced Practice Psychiatric Nurses',
   'EduCode': 3.0},
  {'OnetCode': '29-1299.01',
   'Score': 0.5427063472592409,
   'Rank': 3,
   'Outlook': 'Below Average',
   'AnnualWages': 106230.0,
   'TypicalEducation': "Master's degree",
   'OccupationTitle': 'Naturopathic Physicians',
   'EduCode': 2.0},
  {'OnetCode': '29-1291.00',
   'Score': 0.5307120733164209,
   'Rank': 4,
   'Outlook': 'Average',
   'AnnualWages': 72220.0,
   'TypicalEducation': "Master's degree",
   'OccupationTitle': 'Acupuncturis

#### 2. Send API GET request using custom user answers 

To submit our own request with our own survey answers, we need to imitate the example request we just sent:

In [65]:
with open(os.path.join("data", "skills-submit-api", "example-request.json"), "r") as file:
  example_request = json.load(file)

example_request 

{'SKAValueList': [{'ElementId': '2.C.1.a', 'DataValue': '3.295'},
  {'ElementId': '2.C.4.d', 'DataValue': '3.43'},
  {'ElementId': '2.C.3.d', 'DataValue': '3.1'},
  {'ElementId': '2.C.4.c', 'DataValue': '3.33'},
  {'ElementId': '2.C.1.b', 'DataValue': '3.295'},
  {'ElementId': '2.B.2.i', 'DataValue': '3.31'},
  {'ElementId': '2.C.3.a', 'DataValue': '3.41'},
  {'ElementId': '2.C.1.e', 'DataValue': '3.23'},
  {'ElementId': '2.C.1.c', 'DataValue': '3.175'},
  {'ElementId': '2.C.6', 'DataValue': '3.615'},
  {'ElementId': '2.B.3.j', 'DataValue': '2.375'},
  {'ElementId': '2.C.7.c', 'DataValue': '3.41'},
  {'ElementId': '2.C.7.b', 'DataValue': '3.315'},
  {'ElementId': '1.A.3.c.3', 'DataValue': '2.94'},
  {'ElementId': '2.B.1.e', 'DataValue': '2.44'},
  {'ElementId': '2.B.5.b', 'DataValue': '2.75'},
  {'ElementId': '2.B.5.d', 'DataValue': '2.69'},
  {'ElementId': '2.A.1.e', 'DataValue': '3.06'},
  {'ElementId': '2.C.3.e', 'DataValue': '3.425'},
  {'ElementId': '2.C.5.a', 'DataValue': '3.385'

* `ElementId` - Question identifier (`data-label` attribute in the HTML). Inspected the page source to see which question number each corresponds to (unfortunately aren't all given in order in this example)
* `DataValue`- Beginner, Basic, Skilled, Advanced, Expert. The values which correspond to these ratings change from question to question because they are the unique id tags for the radio button elements in the webpage.   

Each question is listed with a unique identifier that bears no discernible relation to the question number, and each answer level (`DataPoint20`, `DataPoint35`, ... `DataPoint80`) for each question has a globally unique answer format (such that question 37 could be `2.C.7.b` and "Basic" answer could be `3.315`, but question 14 might be `1.A.3.c.3` and "Basic" answer for that could be `2.94`). The questions in the example request above aren't even in order, so we can't translate human-readable question and answers 1-40 based on that assumption. 

In [ ]:
# ## Trying to get questions from HTML page -- Get Skills API endpoint is easier.  
# import requests 
# from bs4 import BeautifulSoup
# url = "https://www.careeronestop.org/Toolkit/Skills/skills-matcher-questions.aspx"

# headers = {}

# response = requests.get(url)
# soup = BeautifulSoup(response.content, 'html.parser')
# soup # Gives 403 Error. Note this is not a request to their API -- I'm just trying to get the survey_questions_questions page's HTML to get the ElementId's and DataValues 

# # Copy-pasted HTML table from inspect element 
# with open("divtableskills.html", "r", encoding="utf-8") as file:
#   skill_table = file.read()

# soup = BeautifulSoup(skill_table, 'html.parser')
# soup # Works, but a bit difficult to extract relevant information (questions are not in order)

The [Get Skills](https://api.careeronestop.org/api-explorer/home/index/SkillsMatcher_GetSkills) API provides the survey questions in JSON (in correct order) with information needed to format our own custom requests:

In [5]:
response = requests.get(url, headers=headers)
response = response.json()
response

{'Skills': [{'ElementId': '2.C.1.a',
   'ElementName': 'Administration and Management',
   'DataPoint20': 1.534,
   'DataPoint35': 2.4145,
   'DataPoint50': 3.295,
   'DataPoint65': 4.1755,
   'DataPoint80': 5.056,
   'AnchorFirst': 'Complete a timesheet',
   'AnchorSecond': '',
   'AnchorThrid': 'Monitor project progress to complete it on time',
   'AnchorFourth': '',
   'AnchorLast': 'Manage a $10m company',
   'Question': 'How much do you know about business planning and leadership?',
   'EasyReadDescription': 'Knowledge of business and management principles involved in strategic planning, resource allocation, human resources modeling, leadership technique, production methods, and coordination of people and resources.'},
  {'ElementId': '2.C.4.d',
   'ElementName': 'Biology',
   'DataPoint20': 1.372,
   'DataPoint35': 2.401,
   'DataPoint50': 3.43,
   'DataPoint65': 4.459,
   'DataPoint80': 5.488,
   'AnchorFirst': 'Care for a pet',
   'AnchorSecond': '',
   'AnchorThrid': 'Investig

In [4]:
import json 

# Loading cached copy
with open(os.path.join("data","get-skills-api","response.json"), "r") as file: # previously extracted with API Explorer
  survey_questions = json.load(file)['Skills']

skill_level_map = { 
  "DataPoint20": "Beginner",
  "DataPoint35": "Basic",
  "DataPoint50": "Skilled",
  "DataPoint65": "Advanced",
  "DataPoint80": "Expert"
}

for q in survey_questions: # Translating Skill level ide
  for k,v in skill_level_map.items():
    q[v] = q.pop(k)

with open(os.path.join("data","get-skills-api","response_translated.json"), "w") as file:
  json.dump(survey_questions, file)

In [5]:
survey_questions

[{'ElementId': '2.C.1.a',
  'ElementName': 'Administration and Management',
  'AnchorFirst': 'Complete a timesheet',
  'AnchorSecond': '',
  'AnchorThrid': 'Monitor project progress to complete it on time',
  'AnchorFourth': '',
  'AnchorLast': 'Manage a $10m company',
  'Question': 'How much do you know about business planning and leadership?',
  'EasyReadDescription': 'Knowledge of business and management principles involved in strategic planning, resource allocation, human resources modeling, leadership technique, production methods, and coordination of people and resources.',
  'Beginner': 1.534,
  'Basic': 2.4145,
  'Skilled': 3.295,
  'Advanced': 4.1755,
  'Expert': 5.056},
 {'ElementId': '2.C.4.d',
  'ElementName': 'Biology',
  'AnchorFirst': 'Care for a pet',
  'AnchorSecond': '',
  'AnchorThrid': 'Investigate effects of pollution on plants',
  'AnchorFourth': '',
  'AnchorLast': 'Identify a new virus',
  'Question': 'How much do you know about plant, animal and cell functions?

Two example functions: 
* One using custom user input for skill ratings
* Another randomly selects skill ratings

In [110]:
def new_survey_request() -> dict:
  """Create new survey request object from user input (will be empty if survey is aborted)"""   
  
  print("Select your skill level")
  print("Use the examples help choose your levels. Think about whether you have done the example activity, or something like it in your own field")
  print("Answer Key: 1 - Beginner; 2 - Basic; 3 - Skilled; 4 - Advanced; 5 - Expert\n")
  print("(Type 'exit' to end this survey)")

  answers = []
  for n,q in enumerate(survey_questions):

    print(f"{n+1}. {q['ElementName']}: {q['Question']}")
    print("1 | 2 | 3 | 4 | 5 ")
    print(f"Beginner: {q['AnchorFirst']}")
    print(f"Skilled: {q['AnchorThrid']}")
    print(f"Expert: {q['AnchorLast']}") 

    x = input()

    if x == "exit":
      return {}
    else: 

      ratings_map = { # Recall that the value for each rating changes from question to question because each is a unique HTML identifier
      
      "1":q['Beginner'], 
      "2":q['Basic'], 
      "3":q['Skilled'],
      "4":q['Advanced'], 
      "5":q['Expert']

      }
      
      answers.append({
        "ElementId": q["ElementId"],
        "DataValue": ratings_map[x], 
      })


  return {"SKAValueList":answers} 

new_request = new_survey_request()

Select your skill level
Use the examples help choose your levels. Think about whether you have done the example activity, or something like it in your own field
Answer Key: 1 - Beginner; 2 - Basic; 3 - Skilled; 4 - Advanced; 5 - Expert

(Type 'exit' to end this survey)
1. Administration and Management: How much do you know about business planning and leadership?
1 | 2 | 3 | 4 | 5 
Beginner: Complete a timesheet
Skilled: Monitor project progress to complete it on time
Expert: Manage a $10m company
2. Biology: How much do you know about plant, animal and cell functions?
1 | 2 | 3 | 4 | 5 
Beginner: Care for a pet
Skilled: Investigate effects of pollution on plants
Expert: Identify a new virus
3. Body Coordination: How well can you coordinate moving your arms, legs, and torso together?
1 | 2 | 3 | 4 | 5 
Beginner: Get in and out of a truck
Skilled: Swim one pool  length, or play a ball sport
Expert: Perform ballet choreography
4. Building and Construction: How much do you know about const

In [121]:
response = requests.post(url, json=new_request, headers=headers)
response = response.json() # This matches the response we generated using the API Explorer
response

{'SKARankList': [{'OnetCode': '29-1221.00',
   'Score': 0.28121313160199796,
   'Rank': 1,
   'Outlook': 'Below Average',
   'AnnualWages': 190350.0,
   'TypicalEducation': 'Doctoral or professional degree',
   'OccupationTitle': 'Pediatricians, General',
   'EduCode': 2.0},
  {'OnetCode': '29-1218.00',
   'Score': 0.272130223767874,
   'Rank': 2,
   'Outlook': 'Below Average',
   'AnnualWages': 999999.0,
   'TypicalEducation': 'Doctoral or professional degree',
   'OccupationTitle': 'Obstetricians and Gynecologists',
   'EduCode': 2.0},
  {'OnetCode': '29-1216.00',
   'Score': 0.262200674428677,
   'Rank': 3,
   'Outlook': 'Below Average',
   'AnnualWages': 214460.0,
   'TypicalEducation': 'Doctoral or professional degree',
   'OccupationTitle': 'General Internal Medicine Physicians',
   'EduCode': 2.0},
  {'OnetCode': '29-1217.00',
   'Score': 0.260975386319619,
   'Rank': 4,
   'Outlook': 'Below Average',
   'AnnualWages': 224260.0,
   'TypicalEducation': 'Doctoral or professional d

In [132]:
import random
def random_survey() -> dict: 
  """Generate a survey response object with random skill ratings for the Submit Skills API"""
  answers = [{"ElementId":q["ElementId"],
      "DataValue":random.choice([q['Beginner'], q['Basic'], q['Skilled'], q['Advanced'], q['Expert']])
    } for q in survey_questions]
    
  return {"SKAValueList":answers} 

rand_request = random_survey()

response = requests.post(url, json=rand_request, headers=headers)
response = response.json() # This matches the response we generated using the API Explorer
response

{'SKARankList': [{'OnetCode': '39-5092.00',
   'Score': 0.443644066481302,
   'Rank': 1,
   'Outlook': 'Bright',
   'AnnualWages': 31130.0,
   'TypicalEducation': 'Postsecondary non-degree award',
   'OccupationTitle': 'Manicurists and Pedicurists',
   'EduCode': 6.0},
  {'OnetCode': '19-3091.00',
   'Score': 0.44286812419121596,
   'Rank': 2,
   'Outlook': 'Average',
   'AnnualWages': 63940.0,
   'TypicalEducation': "Master's degree",
   'OccupationTitle': 'Anthropologists and Archeologists',
   'EduCode': 2.0},
  {'OnetCode': '45-2041.00',
   'Score': 0.43716886592564197,
   'Rank': 3,
   'Outlook': 'Below Average',
   'AnnualWages': 32550.0,
   'TypicalEducation': 'No formal educational credential',
   'OccupationTitle': 'Graders and Sorters, Agricultural Products',
   'EduCode': 9.0},
  {'OnetCode': '31-1131.00',
   'Score': 0.43314916662403297,
   'Rank': 4,
   'Outlook': 'Bright',
   'AnnualWages': 35760.0,
   'TypicalEducation': 'Postsecondary non-degree award',
   'OccupationTi

The API states you can filter out suggested jobs from the response by setting an education level restriction in the parameters, but it doesn't do anything when I provide it (example below). An easy workaround is to omit that parameter, get the entire response object of suggested jobs, and filter by education level afterwards.

In [140]:
headers_associates = {
  "Authorization": f"Bearer {api_key['token-key']}", 
  "eduFilterValue":"Associate's degree"
}
response = requests.post(url, json=rand_request, headers=headers_associates)
response = response.json() # This matches the response we generated using the API Explorer
response

{'SKARankList': [{'OnetCode': '39-5092.00',
   'Score': 0.443644066481302,
   'Rank': 1,
   'Outlook': 'Bright',
   'AnnualWages': 31130.0,
   'TypicalEducation': 'Postsecondary non-degree award',
   'OccupationTitle': 'Manicurists and Pedicurists',
   'EduCode': 6.0},
  {'OnetCode': '19-3091.00',
   'Score': 0.44286812419121596,
   'Rank': 2,
   'Outlook': 'Average',
   'AnnualWages': 63940.0,
   'TypicalEducation': "Master's degree",
   'OccupationTitle': 'Anthropologists and Archeologists',
   'EduCode': 2.0},
  {'OnetCode': '45-2041.00',
   'Score': 0.43716886592564197,
   'Rank': 3,
   'Outlook': 'Below Average',
   'AnnualWages': 32550.0,
   'TypicalEducation': 'No formal educational credential',
   'OccupationTitle': 'Graders and Sorters, Agricultural Products',
   'EduCode': 9.0},
  {'OnetCode': '31-1131.00',
   'Score': 0.43314916662403297,
   'Rank': 4,
   'Outlook': 'Bright',
   'AnnualWages': 35760.0,
   'TypicalEducation': 'Postsecondary non-degree award',
   'OccupationTi